### PR TITLE
Adding MED format, second submission

### PR DESCRIPTION
**Pull request by dancrepeau**
Hello,

A month or so ago, my colleague Matt Stead and I pushed a fork with changes to support the MED format (medformat.org).  This new pull request is for a version that's a bit more detailed: it finds discontinuities and creates separate Neo segments, and it groups channels together that have the same sampling frequencies.  It also reads Note and Neuralynx type annotations, and has two time basis options, zero-based and original time. 

Please let me know if there are questions/comments/criticisms!  Thank you,
Dan

### CODE REVIEW

File: neo/test/iotest/test_medio.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+          
+        r.close()
+
+    def test_read_block(self):
+  
+        r = MedIO(self.dirname, self.password)
+        bl = r.read_block(lazy=True)
+        self.assertTrue(bl.annotations)
+        
+        seg = bl.segments[0]
```
Comment (JuliaSprenger) [inline neo/test/iotest/test_medio.py:85]: Depending on your test file the IO should generate more than 1 segment. If there's more segments, could you also check those?

**Review by JuliaSprenger — COMMENTED** [review]
Hi @dancrepeau!
Thanks for rewriting the IO! I think the segment & stream concept is not implemented consistently yet. I left some comments in the code. Tell me if you need more details on this. I think a test file with more than one segment (by design or discontinuous data) would be beneficial here to implement this.  Based on the tests you wrote I guess the current test file is a single segment only, right?

**Comment by dancrepeau** [comment]
Hi @JuliaSprenger!

Thank you for reviewing the code.  I will go through your comments thoroughly in the coming days.  I will comment right away that we are indeed testing against a recording that has discontinuous ranges, specifically "test.medd".  Starting on line 76 of test_medrawio.py, we are verifying that it found 3 segments, and there are two unique streams in each segment (6 streams total).

If streams are supposed to be shared across all segments, then I will have to re-work the code.  I'm using the convention that a continuous stretch of data is a segment, and within that, a stream is a grouping of channels that have the same sampling frequency.  Please correct me if I have those ideas wrong.  

My notion of a "template" (which admittedly isn't clear) is just a grouping of channels that have the same sampling frequency.  However, there can be lots of discontinuitites within a recording, so I am multiplying the templates by continuous ranges, and assuming those are streams.  I am assuming a stream can't have gaps in it.

Thanks again for all the help!
-Dan

**Comment by JuliaSprenger** [comment]
Hi @dancrepeau,
yes, I think the segment & stream design might need to be adjusted to fit the Neo & RawIO structures.
In general a Neo segment is intended to contain data that shares a common clock. So for many IOs multiple segments are generated when either the recording was paused and restarted or the recording system didn't manage to write all of the continuous data, but lost some samples at a certain time point.

In the RawIO design a stream is independently of segments describing different `input channels`, characterised e.g. for MED by their different sampling rates. When loading a dataset Neo creates one AnalogSignal per Segment per Stream. This also implies that all channels (recording traces) in a stream have to contain the same number of samples within a segment, to be able to represent the data of that AnalogSignal as a numpy array.

I hope that helps restructuring, let me know if you are still missing some information.

**Comment by dancrepeau** [comment]
Hi @JuliaSprenger,

Thank you for the clarifications.  The terminology here can become very confusing.  In MED, segments and continuous sections are independent concepts, as segments are primarily used to break recordings up into smaller files, but can also be used to indicate a new experiment within a recording.

I might suggest some changes to documentation to make these Neo concepts easier for newcomers to digest (and feel free to disregard me here, as I am jus trying to be helpful).  The Neo core documentation (https://neo.readthedocs.io/en/stable/core.html) only shows a higher level view; indeed the word "stream" doesn't appear anywhere on this page.  I found the file "baserawio.py" to be helpful in understanding the raw io concepts, however I came to some wrong conclusions.  For example, the function:      

def get_signal_t_start(self, block_index, seg_index, stream_index=None):
        """
        Retrieve the t_start of a single section of the channels in a stream.
        :param block_index:
        :param seg_index:
        :param stream_index:
        :return: start time of section
        """
refers to a 'section' of a stream, yet 'section' is not a parameter to the function.  I believe the correct interpretation is that 'segment' and 'section' are interchangeable (in this context).  I came to the conclusion that 'section' is implied by the stream itself, and therefore there was only one start time to a stream.

Again, I appreciate the clarifications!
-Dan

File: neo/rawio/medrawio.py [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+    to be a new segment.  Channels are grouped by sampling frequency, to
+    create streams.  So every continuous stretch of data with the same
+    sampling frequency is a stream.
+    
+    Timestamps generated are referenced to the beginning of the session,
+    with the beginning of the session being timestamp zero.  If UTC timestamps
+    are desired, then the keep_original_times flag in the constructor can be
+    set to True (it defaults to False) and the timestamps used in the object
+    will be seconds, reference to midnight 1 Jan 1970 (assuming that that
+    data is available in the MED data session).
```
[the remark this answers is not in the document]
Comment (dancrepeau) [reply neo/rawio/medrawio.py:32]: Hello, just going through these comments again.  No, there is no performance difference here, just personal preference.

File: neo/rawio/medrawio.py [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+
+    def _rescale_event_timestamp(self, event_timestamps, dtype, event_channel_index):
+        return np.asarray(event_timestamps, dtype=dtype)
+
+    def _rescale_epoch_duration(self, raw_duration, dtype, event_channel_index):
+        return np.asarray(raw_duration, dtype=dtype)
+        
+    def __del__(self):
+        try:
+            # Important to make sure the session is closed, since the MED library only allows
```
[the remark this answers is not in the document]
Comment (dancrepeau) [reply neo/rawio/medrawio.py:336]: In principle the files can be opened many times, however the dhn_med_py python library is designed for the use-case that a single session will be opened at a time.  Within the library, channel reading is all done in parallel.

[the remark this answers is not in the document]
Comment (JuliaSprenger) [reply neo/rawio/medrawio.py:336]: Ok, as long as it is possible to open the same file multiple times in parallel this should be compatible.

**Pull request closed without merging**